src/parallel_prover.rs

```diff
@@ -336,48 +334,151 @@ where
-        &right.r_W_secondary,
-        &right.l_u_secondary,
-        &right.l_w_secondary,
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &pp.r1cs_shape_secondary,
+      &right.r_U_secondary,
+      &right.r_W_secondary,
+      &right.l_u_secondary,
+      &right.l_w_secondary,
     )?;
     let (nifs_secondary, (U_secondary, W_secondary)) = NIFS::prove(
-        &pp.ck_secondary,
-        &pp.ro_consts_secondary,
-        &pp.r1cs_shape_secondary,
-        &left_U_secondary,
-        &left_W_secondary,
-        &right_U_secondary,
-        &right_W_secondary,
+      &pp.ck_secondary,
+      &pp.ro_consts_secondary,
+      &pp.r1cs_shape_secondary,
+      &left_U_secondary,
+      &left_W_secondary,
+      &right_U_secondary,
+      &right_W_secondary,
     )?;
 
-
     return Err(NovaError::InvalidStepOutputLength);
-  } 
-}
\ No newline at end of file
+  }
+}
+
+/// Structure for parallelization
+#[derive(Debug, Clone)]
+pub struct ParallelSNARK<G1, G2, C1, C2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+{
+  nodes: Vec<Vec<NovaTreeNode<G1, G2, C1, C2>>>,
+}
+
+/// Implementation for parallelization SNARK
+impl<G1, G2, C1, C2> ParallelSNARK<G1, G2, C1, C2>
+where
+  G1: Group<Base = <G2 as Group>::Scalar>,
+  G2: Group<Base = <G1 as Group>::Scalar>,
+  C1: StepCircuit<G1::Scalar>,
+  C2: StepCircuit<G2::Scalar>,
+{
+  /// Create a new instance of parallel SNARK
+  pub fn new(
+    pp: &PublicParams<G1, G2, C1, C2>,
+    steps: usize,
+    z0_primary: Vec<G1::Scalar>,
+    z0_secondary: Vec<G2::Scalar>,
+    c_primary: C1,
+    c_secondary: C2,
+  ) -> Self {
+    let mut zi = Vec::<(usize, Vec<G1::Scalar>, Vec<G2::Scalar>)>::new();
+    zi.push((0, z0_primary.clone(), z0_secondary.clone()));
+    for i in 1..steps {
+      let (index, prev_primary, prev_secondary) = &zi[i - 1];
+      zi.push((
+        i,
+        c_primary.output(&prev_primary),
+        c_secondary.output(&prev_secondary),
+      ));
+    }
+    // Do calculate argumeted inputs in parallel
+    let leafs_vec = zi
+      .par_iter()
+      .map(|item| {
+        let (index, zi_primary, zi_secondary) = item;
+        NovaTreeNode::new(
+          &pp,
+          c_primary.clone(),
+          c_secondary.clone(),
+          *index,
+          z0_primary.clone(),
+          (*zi_primary).clone(),
+          z0_secondary.clone(),
+          (*zi_secondary).clone(),
+        )
+        .expect("Unable to create basic leaf")
+      })
+      .collect();
+    // Create a new parallel prover wit basic leafs
+    Self {
+      nodes: vec![leafs_vec],
+    }
+  }
+
+  /// Perform the proving in parallel
+  pub fn prove(&mut self, pp: &PublicParams<G1, G2, C1, C2>, c_primary: &C1, c_secondary: &C2) {
+    // Calculate the max height of the tree
+    // ⌈log2(n) + 1⌉
+    let max_height = ((self.nodes[0].len() as f64).log2().ceil() + 1f64) as usize;
```

Comment:
I think the height is actually log_2(n), because in our cause the base of the tree is half the length of the steps.

Reply:
- `log_2(n)` is incorrect for `5` leafs since it's equal to `2`
- `ceil(log2(n)) + 1` is correct for `5` leafs since it's equal to `4`